### DISCUSSION TITLE
Documentation, custom rules and StringSegment.Create(...);

### DISCUSSION THREAD

**Discussion by FatBobSlim** [MediatedCommunications/Mime-Detective#41] [Q&A]
Hi we're trying to add a few custom rules into the default set of rules mime-detective comes with.

The example from the project readme shows an example of how to do this, and in particular how to use sets of StringSegment.Create(...), and PrefixSegment.Create(...) statements to define any one particular custom definition.

When I try to implement similar though, it doesn't seem to work as I'd expect.
From the example: 

"StringSegment.Create("MAGIC"), //anywhere in the file, expect "MAGIC" (exact case)."

However, when I'm trying this, it doesn't match the "exact case" of the text, it matches any case of the text, so "magic", "Magic" etc all match from this rule.
Also when I try to create a segment using StringSegment.Create("< someTag >"), to match "< someTag >" anywhere in the file, it doesn't find that text. If I assume the text has to be uppercase in the Create call, StringSegment.Create("< SOMETAG >"), it still doesn't match that text of any case anywhere in the file. (Ignore the spaces around the < and > characters, it doesn't like them when I try to post without the spaces)
If I take the "<" and ">" characters out of the string, so use instead StringSegment.Create("SOMETAG"), it then matches any casing of "sometag" in the file.

So this would suggest, you have to specify the text in uppercase, and then it will match any case of the text. (?)
It may also suggest that symbols other than alphanumerics cannot be used. (?)

It's worth mentioning that PrefixSegment does seem to work as you'd hope when given " < " symbols etc, when looking for text at fixed positions. So it would seem StringSegment works differently.

I have also tried the other overload on StringSegment.Create that takes an IEnumerable<byte>, and converting the text "< SOMETAG >" (without the spaces) to bytes and then sending that in. It also doesn't match.

I could be adding 2 + 2 and getting 5 here (!), or just doing something daft of course, but hopefully the examples above illustrate what I'm unclear on.

I wanted to know if there is any more detailed documentation for how to set up custom rules, and/or some explanation of how this sort of thing should be done please?

**Comment by TonyValenti** [comment]
I'm OOO for two weeks traveling internationally. I'll have to take a look at this when I get back.

**Reply by FatBobSlim** [reply]
> I'm OOO for two weeks traveling internationally. I'll have to take a look at this when I get back.

Ok, many thanks.

**Comment by FatBobSlim** [comment]
After looking in the source code for this I think I can answer a lot of what I was asking.

What I think I know (!) :-) .... 

PrefixSegments take a hex string of bytes and match against the bytes provided, so are implicitly case sensitive. You always provide an offset index to state where the bytes should be found in the file for a match.

StringSegments take a string to look for or an array of byte. However, a conversion is performed when matching string segments, to treat all the file contents as uppercase. So all strings (or bytes) provided to a StringSegment must be provided as an uppercase representation in order to match. If you provide anything lowercase it won't match.
There also seems to be a list of symbols in the StringSegmentExtrator file: " -+_.$(){}~*%\0".
So it looks like StringSegment strings or bytes provided can only contain these special characters outside the usual alphanumerics, which is why it fails to match when I try a string with a '<' or '>', or '!' etc.

static StringSegmentExtrator() {
            var Valid = ""
                + "abcdefghijklmnopqrstuvwxyz"
                + "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
                + "0123456789"
                + " -+_.$(){}~*%\0"
                ;

So once you're back and have a moment to take a look, if you could confirm whether the above is accurate, particularly regarding whether there are restrictions on the symbols you can use for StringSegments (but not PrefixSegments), and why that would be, that would be really helpful? Thanks in advance.
